Support timing function for animation. Fix #12 #13

phimage opened the pull request:
Re pull #456
Timing function enum added `TiminFunctionType`
  - Add it to `Animatable` objects and `AnimationConfiguration`
  - Use this enum when possible instead of `CAMediaTimingFunction`

TODO
- [x] add more type in enum (see `extension CAMediaTimingFunction`)
- [x] add more type in enum  see `CAMediaTimingFunction(` contructors in code
- [x] add custom type with control points in enum ?
- [x] manage spring animation or not, animation default value
- [x] Add an example
- [x] Update CHANGELOG
- [x] Documentation

phimage committed:
Revert "Revert "Support timing function for animation. Fix #12" #no-public-changes"

e-marchand committed:
In demo app select previously selected timing function

e-marchand committed:
Fix moteBy and moveTo animation
With TimingFunction position is animated using centre point

e-marchand committed:
Update change log

e-marchand committed:
Fix swiftlint warning about space

tbaranes commented:
@phimage Sorry it takes me time to review it, it's quite big and a lot to digest! 
I have already a first comment: can you update the docs? I think you will the one to update is `Documentation/APIs.md`

phimage commented:
No problem!

API doc has not been reverted https://github.com/IBAnimatable/IBAnimatable/blob/master/Documentation/APIs.md
I add `timingFunction`  in "Animatable protocol" section.
If you think I can add something elsewhere, I will do it :)

I see x, y, repeat.. attributes in the same section, maybe old parameters, now directly  in animationType

tbaranes commented on IBAnimatable/TimingFunctionType.swift:
```diff
@@ -0,0 +1,355 @@
+      let c2 = (params[2].toFloat() ?? 0, params[3].toFloat() ?? 0)
+      self = .custom(c1, c2)
+
+    case "easeinsine":
```
Can you add a comment as the ones above to keep the same structure as the enum?

tbaranes commented on IBAnimatable/TimingFunctionType.swift:
```diff
@@ -0,0 +1,355 @@
+  case easeInExpo, easeOutExpo, easeInOutExpo
+  case easeInCirc, easeOutCirc, easeInOutCirc
+  case easeInBack, easeOutBack, easeInOutBack
+
```
Typo: there's one newline to remove

tbaranes commented on IBAnimatable/TimingFunctionType.swift:
```diff
@@ -0,0 +1,355 @@
+    case "easeinoutback":
+      self = .easeInOutBack
+
+    default:
```
Same here, I know it's already explicit, but I think it's better to keep the same structure as the enums. It helps to follow

tbaranes commented on IBAnimatable/TimingFunctionType.swift:
```diff
@@ -0,0 +1,355 @@
+extension TimingFunctionType {
+
+  public var caType: CAMediaTimingFunction {
+    switch self {
```
Same as above, if you can add the comments to separate the main groups

tbaranes replied:
Github didn't see as fixed, but it's fixed!

tbaranes commented on IBAnimatable/TimingFunctionType.swift:
```diff
@@ -0,0 +1,355 @@
+
+}
+
+@available(iOSApplicationExtension 10.0, *)
```
`iOSApplicationExtension`, did you mean `iOS`?

tbaranes commented on IBAnimatable/TimingFunctionType.swift:
```diff
@@ -0,0 +1,355 @@
+
+}
+
+internal extension CAAnimation {
```
Is there a reason to explicitly make it `internal`?

phimage replied:
It's a way for me to mark dangerous extension
I want to use it internally but never let it public, or someone else does it.

With `internal` explicitly marked, people ask themselves why and edit carefully

tbaranes commented on IBAnimatable/TimingFunctionType.swift:
```diff
@@ -0,0 +1,355 @@
+
+internal extension CAAnimation {
+
+  // conveniant setter, getter not implemented
```
Typo:  `conveniant` -> `convenient`

tbaranes commented on IBAnimatableApp/AnimationsTimingFunctionViewController.swift:
```diff
@@ -0,0 +1,233 @@
+  func timingFunctionSelected(_ timingFunction: TimingFunctionType)
+}
+
+class AnimationsTimingFunctionViewController: UIViewController {
```
final

tbaranes commented on IBAnimatableApp/AnimationsViewController.swift:
```diff
@@ -12,13 +12,14 @@ private let fadeWayParams = ParamType(fromEnum: AnimationType.FadeWay.self)
+private let numberParam = ParamType.number(min: -50, max: 200, interval: 10, ascending: true, unit: "")
 private let repeatCountParam = ParamType.number(min: 1, max: 10, interval: 1, ascending: true, unit:"")
 
 class AnimationsViewController: UIViewController {
```
final

phimage replied:
I think do a PR to set all `final` it's a better choice
Because here it's not a new code here. But I copy this code to make my new controller :-)

tbaranes replied:
Yeah, I know that we didn't add the final at the beginning of this project, we should have this in our roadmap: add all the missing finals 💥

tbaranes reviewed (changes requested):
That's a HUGE work, thanks for doing this, it's an awesome feature.
I let a few comments, mainly typos / minor stuff, it looks pretty good to me 👍 

Can you add the docs update once more? I think that's enough.

tbaranes commented:
Just tested the example, great one. A few feedbacks:

- At first, I though that the timing function will update the animation. Since it doesn't, you should either do it, or refactoring makes a subsections to "Animations" to have two possibilities: "View Animations", and "Timing functions"
- The image in the navigation bar isn't explicit enough (I didn't get it at first), if you choose option 1 from the previous point, we should replace the image by a text, or move the button in a place even more visible / explicit
- Since some `timingFunctions` are quite long, they are truncated in the picker. Can we add the selected name somewhere? Below the graph for example. I think we have this issue in others picker as well, time to take care of this by starting with this one? 😬 
- I know there's a few incoherence about this in the playground: should we have a "back" text and no title, or no "back" text and a title? By using "TimingFunctions", I would go for option 2 to make it more explicit, but it's up to you. Some day it would be great to uniformise that behaviour

phimage commented:
The animation is updated in the previous view, `TimingFunctionPickDelegate` do it

But not all animation are compatible ie. Spring animation are not compatible
In the same way velocity, dumping are only for spring animation and update it will not affect a lot of animations
Maybe a documentation file with a list of animation with two columns: Spring animation and timing function animation(or cubic animation) will be necessary, to check option compatibility

I edit moveTo and moveBy to be compliant with the two modes.

---
Yes, the image is not explicit
I take an available image, a timer. I think about putting a timing function graph but failed to create it using my TimingFunctionView in navigation bar items.

phimage committed:
In demo app timing function button is enabled is animation is configurable by timing function
Redesign back button and title for timing function picker controller
Typo, comment on TimingFunctiontType

phimage commented:
So new change added

- back button and navigation title for the timing function picker
- icon changed to a Gear, and the button is enabled only if the selected animation is configurable using the timing function

Added on AnimationType two compute boolean variables to know is the animation use velocity and damping or use timing function
This method could help for future animation documentations

SD10 commented:
@phimage this is great work! I may be able to leverage this approach to do more cleanup or adopt a similar strategy. love it!

tbaranes reviewed (changes requested):
@phimage Awesome! I agree with you that a docs about timing function compatibility would be great, but we can merged that one first. So, just the missing entry in the docs and that's ready to go 👍

phimage commented:
For the doc, as I say, I do already something in main branch, see Animatable protocol
https://github.com/IBAnimatable/IBAnimatable/blob/master/Documentation/APIs.md

phimage committed:
Merge branch 'master' into revert-473-revert-456-feature/timingfunction

tbaranes reviewed (approved):
My bad, I misunderstood 😶

JakeLin commented on IBAnimatable/ActivityIndicatorAnimationOrbit.swift:
```diff
@@ -118,17 +118,16 @@ private extension ActivityIndicatorAnimationOrbit {
-    let timingFunction = CAMediaTimingFunction(controlPoints: 0.19, 1, 0.22, 1)
     opacityAnimation.keyTimes = [0, 0.45, 1]
-    ring1ScaleAnimation.timingFunctions = [CAMediaTimingFunction(name: kCAMediaTimingFunctionLinear), timingFunction]
+    opacityAnimation.timingFunctionsType = [.linear, .easeOutExpo]
```
good to encapsulate to an extension for timing function.

JakeLin commented on IBAnimatable/Animatable.swift:
```diff
@@ -44,9 +44,14 @@ public protocol Animatable: class {
 }
 
-public extension Animatable where Self: UIView {
+public extension Animatable {
```
I hope Swift can support the syntax like `public extension Animatable where Self: UIView or UIBarItem` later.

JakeLin commented on IBAnimatable/Animatable.swift:
```diff
@@ -186,22 +193,45 @@ fileprivate extension Animatable where Self: UIView {
+      path.move(to: position)
+      path.addLine(to: CGPoint(x: xToMove, y: yToMove))
+
+      animatePosition(path: path, configuration: configuration, completion: completion)
```
Cool, before iOS 10, only Core Animation can apply custom timing.

JakeLin commented on IBAnimatable/Animatable.swift:
```diff
@@ -589,3 +655,42 @@ public extension Animatable where Self: UIBarItem {
+  }
+
+  /// This animation use timing function parameter.
+  public var isCubic: Bool {
```
Should it call something like `doesSupportTimingFunction`? I am not good at naming though 😂

JakeLin commented on IBAnimatable/AnimatableBarButtonItem.swift:
```diff
@@ -41,6 +41,12 @@ open class AnimatableBarButtonItem: UIBarButtonItem, BarButtonItemDesignable, An
   @IBInspectable open var damping: CGFloat = CGFloat.nan
   @IBInspectable open var velocity: CGFloat = CGFloat.nan
   @IBInspectable open var force: CGFloat = CGFloat.nan
+  @IBInspectable open var _timingFunction: String = "" {
```
This one should not be `open`. It only is used in Interface Builder. Please have a look at the comment around https://github.com/IBAnimatable/IBAnimatable/issues/218#issuecomment-230324948 . We don't want the `_` properties be used in programmatical APIs.

JakeLin commented on IBAnimatable/AnimatedTransitioning.swift:
```diff
@@ -47,8 +47,7 @@ public extension AnimatedTransitioning {
       transition.duration = self.transitionDuration(using: transitionContext)
-      // Use `EaseOutQubic` for system built-in transition animations. Thanks to @lexrus
-      transition.timingFunction = CAMediaTimingFunction(controlPoints: 0.215, 0.61, 0.355, 1)
+      transition.timingFunctionType = .easeOutCubic
```
Great 🎉

JakeLin commented on IBAnimatable/TimingFunctionType.swift:
```diff
@@ -0,0 +1,360 @@
+  case custom((x: Float, y: Float), (x: Float, y: Float))
+  case spring(damping: Float)
+
+  // from http://easings.net/
```
👍

JakeLin commented on IBAnimatable/TimingFunctionType.swift:
```diff
@@ -0,0 +1,360 @@
+
+@available(iOS 10.0, *)
+extension TimingFunctionType {
+  var cubicTimingParameters: UICubicTimingParameters {
```
Supper cool to support `UIPropertyAnimator` API.

JakeLin reviewed (approved):
All good, great work 👍👍👍. Look forward to using it.

JakeLin committed:
Merge branch 'master' into revert-473-revert-456-feature/timingfunction

JakeLin commented:
@phimage thanks for such a great PR. I think PR is ready to go. Please have a look at this comment https://github.com/IBAnimatable/IBAnimatable/pull/478#discussion_r116907694 for the `_timingFunction`. It should be `internal` like the other `_` properties.

e-marchand committed:
Set internal _timingFunction variable

JakeLin commented:
@phimage Well done 🎉🎉🎉🎉🎉🎉
